### PR TITLE
session: add more information for retry error

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
It is hard to know why transaction fail when `tidb_disable_txn_auto_retry` is off. Printing some important variable value is helpful for us to locate problems.

### What is changed and how it works?
1. Printing variable value associated with transaction retry when `isTxnRetryableError` is true
2. Using `GenWithStackByArgs` for  `RetryableError` rather  than `FastGenByArgs`

### Check List 

Tests 

 - Unit test

Code changes

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

Release note

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -475,6 +475,9 @@ func (s *session) doCommitWithRetry(ctx context.Context) error {
 			txnSizeRate := float64(txnSize) / float64(kv.TxnTotalSizeLimit)
 			maxRetryCount := commitRetryLimit - int64(float64(commitRetryLimit-1)*txnSizeRate)
 			err = s.retry(ctx, uint(maxRetryCount))
+		} else if s.isTxnRetryableError(err)  {
```
Comment (jackysp) [inline session/session.go]: I think there is no need to check `isTxnRetryableError` again.

File: session/session.go [hunk 2]
```diff
@@ -475,7 +475,15 @@ func (s *session) doCommitWithRetry(ctx context.Context) error {
 			txnSizeRate := float64(txnSize) / float64(kv.TxnTotalSizeLimit)
 			maxRetryCount := commitRetryLimit - int64(float64(commitRetryLimit-1)*txnSizeRate)
 			err = s.retry(ctx, uint(maxRetryCount))
+		} else {
+			logutil.Logger(ctx).Warn("can not retry txn",
+				zap.String("label", s.getSQLLabel()),
+				zap.Error(err),
+				zap.Bool("isBatchInsert", s.sessionVars.BatchInsert),
+				zap.Bool("isPessimistic", isPessimistic),
+				zap.Int64("commitRetryLimit", commitRetryLimit))
```
Comment (jackysp) [inline session/session.go]: Plz add the value of `tidb_disable_txn_auto_retry` and `tidb_retry_limit` and `InRestrictedSQL`. `commitRetryLimit` is not very useful for debug.

Comment (imtbkcat) [reply session/session.go]: @jackysp PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**